### PR TITLE
Implement SafeStatusHandle

### PR DESCRIPTION
**Pull request by sharwell**
This pull request converts use of `TF_Status*` to a proper `SafeHandle`. This is the first of a rather long sequence of changes to improve the overall reliability of TensorFlow.NET. This particular type was relatively easy to convert, so I started with it. I am happy to answer any questions you may have.

⚠️ This pull request is separated into one commit for each distinct operation. It is highly recommended that the review be commit-by-commit, and the merge (if accepted) preserve the commits in this pull request.

* The first commit introduces `SafeTensorflowHandle` with some extension methods.
* The second commit introduces `SafeStatusHandle` and uses it where appropriate.

📝 I already implemented conversions for many additional resources locally. I'm waiting for feedback on this proposal before submitting each of those individually.

### CODE REVIEW

File: test/TensorFlowNET.UnitTest/Eager/CApi.Eager.cs [hunk 1]
```diff
@@ -17,25 +17,23 @@ IntPtr TestMatrixTensorHandle()
...
+            using var status = TF_NewStatus();
 
             var op = TFE_NewOp(ctx, "MatMul", status);
             CHECK_EQ(TF_OK, TF_GetCode(status), TF_Message(status));
             TFE_OpAddInput(op, a, status);
             CHECK_EQ(TF_OK, TF_GetCode(status), TF_Message(status));
             TFE_OpAddInput(op, b, status);
             CHECK_EQ(TF_OK, TF_GetCode(status), TF_Message(status));
-            TF_DeleteStatus(status);
             TFE_OpSetAttrType(op, "T", TFE_TensorHandleDataType(a));
```
Comment (sharwell) [inline test/TensorFlowNET.UnitTest/Eager/CApi.Eager.cs]: 📝 This one changed the order of operations, but it seemed acceptable for two reasons:

1. It didn't seem to matter which order these two lines appear in, as the second does not use `status`
1. This is test code, so it's not risking a subtle change to shipping code to switch the order here

A few other tests in the same file have the same situation.

File: src/TensorFlowNET.Core/Status/Status.cs [hunk 2]
```diff
@@ -60,19 +72,17 @@ public void Check(bool throwException = false)
...
+                var message = Message;
+                Console.WriteLine(message);
                 if (throwException)
-                    throw new TensorflowException(Message);
+                    throw new TensorflowException(message);
             }
         }
 
-        public static implicit operator IntPtr(Status status)
-            => status._handle;
```
Comment (sharwell) [inline src/TensorFlowNET.Core/Status/Status.cs]: 📝 The implicit operators are _extremely_ problematic, and make the type impossible to use reliably. Explicit operators are equally impossible to use correctly, so I'm removing all of these operators as I work my way through the various types of native resources.

File: src/TensorFlowNET.Core/Status/Status.cs [hunk 3]
```diff
@@ -25,26 +26,37 @@ namespace Tensorflow
...
         /// </summary>
-        public string Message => c_api.StringPiece(TF_Message(_handle));
+        public string Message
+        {
+            get
+            {
+                using (Handle.Lease())
+                {
+                    return StringPiece(TF_Message(Handle));
+                }
```
Comment (sharwell) [inline src/TensorFlowNET.Core/Status/Status.cs]: The .NET runtime will automatically acquire a lease when the `SafeHandle` is passed to `TF_Message`. However, the return value from `TF_Message` is an `IntPtr` which is also tied to the lifetime of the `SafeStatusHandle`. By wrapping the use of this pointer in a lease, we ensure the `SafeStatusHandle` is not finalized prior to the completion of the call to `StringPiece`.

📝 It's not normally desirable to use `Handle.Lease()` in coding patterns, but for cases where we _can't_ automatically have things work through proper P/Invoke signatures, it allows the code to still be reliable.

File: src/TensorFlowNET.Core/Status/Status.cs [hunk 4]
```diff
@@ -25,26 +26,37 @@ namespace Tensorflow
     /// TF_Status holds error information. It either has an OK code, or
     /// else an error code with an associated error message.
     /// </summary>
-    public class Status : DisposableObject
```
Comment (sharwell) [inline src/TensorFlowNET.Core/Status/Status.cs]: 📝 `DisposableObject` follows legacy resource management patterns that are no longer recommended for use. As types are migrated to new patterns, they will no longer use `DisposableObject`, and eventually the type can simply be deleted.

File: src/TensorFlowNET.Core/Status/Status.cs (hunk 2 above)
Comment (Oceania2018) [reply src/TensorFlowNET.Core/Status/Status.cs]: OK, I don't even know this tricky.

**Comment by Oceania2018** [comment]
Please make this test pass. Below is my master branch.

image

I can't even run (memory crashed) in your branch.

image

**Comment by sharwell** [comment]
@Oceania2018 I will make the tests pass locally. Is there anything we can do to help make the tests pass in CI?

**Comment by sharwell** [comment]
@Oceania2018 I'm getting this error:

```
    Test method TensorFlowNET.UnitTest.ConstantTest.OnesConst threw exception: 
    System.EntryPointNotFoundException: Unable to find an entry point named 'TFE_OpReset' in DLL 'tensorflow'.
```

**Comment by Oceania2018** [comment]
I will upload a test TensorFlow binary to NuGet. The customized package added more TensorFlow C functions.

**Comment by sharwell** [comment]
I'm not sure there's much I can do here prior to a proper dependency getting published. Right now the source refers to a non-existent library:

https://github.com/SciSharp/TensorFlow.NET/blob/784b1ebee9bbced565e1b8729e110451c30f8bc4/src/TensorFlowNET.Core/APIs/c_api.cs#L46

**Comment by Oceania2018** [comment]
@sharwell Sorry about this debug code. `master` branch is still not stable. Will let you know once we published stable `0.20.0`.

**Comment by Oceania2018** [comment]
Sync code, should be working.

**Comment by sharwell** [comment]
@Oceania2018 I updated my branch, but GitHub won't allow me to reopen this issue.

**Comment by sharwell** [comment]
@Oceania2018 If you have questions related to how/why I implemented any part of this the way I did, please let me know. I'm happy to discuss this further.

**Comment by Oceania2018** [comment]
@sharwell Can you sync with master branch and fix the compile issue. The unit tests will be failed as expected, it's OK as long as it can be compiled.

image

**Comment by sharwell** [comment]
@Oceania2018 done!

**Comment by Oceania2018** [comment]
@sharwell It still shows conflicting files.

**Comment by sharwell** [comment]
> It still shows conflicting files.

You pushed to master after my update. 😆 

I'll update again...

**Comment by sharwell** [comment]
@Oceania2018 updated again!

**Comment by Oceania2018** [comment]
@sharwell Can you sync with the upstream ?

image

**Comment by sharwell** [comment]
> Can you sync with the upstream ?

Best to just ignore branches in my fork that aren't part of a pull request. I update directly against this repository.

**Comment by sharwell** [comment]
@Oceania2018 If you ever have questions about how to incorporate the patterns seen here with new APIs/code you are integrating with, please feel free to ask (an `@sharwell` mention is a good way)!

**Pull request merged**